features/256-auth-gear-interaction-overview-in-next/api.md

```diff
@@ -0,0 +1,441 @@
+| `metadata` | `before_metadata_changed_sync(user, original_user, context)`<br/>`before_metadata_changed(user, original_user, context)`<br/>`after_metadata_changed_sync(user, original_user, context)`<br/>`after_metadata_changed(user, original_user, context)` |
+| `user` | `before_user_changed_sync(user, original_user, context)`<br/>`before_user_changed(user, original_user, context)`<br/>`after_user_changed_sync(user, original_user, context)`<br/>`after_user_changed(user, original_user, context)` |
+
+1. Hooks listed presented here are based on a assumption that a developer could use `content.req.path` to know the reason of certain user auth data changed.
+
+   For example, a user's password changed could be due to:
+
+   1. admin reset a user's password.
+   2. a user changes password by himself.
+   3. a user forgot password, and reset password via email form.
+
+   They are corresponding to following `context.req.path`:
+
+   1. `/reset_password`
+   2. `/change_passowrd`
+   3. `/forgot_password/reset_password`
+
+2. `user_changed` hooks are useful when a developer isn't sure which hook to implement, or it can be a hook to centralize a business logic that may be crossed over many hooks.
+
+3. And to avoid spiral request loop, it is forbidden to send request to auth gear in hooked Cloud Function.
+
+4. `context` should contain following information:
+
+   1. `context.user`: user who triggers the hook, ex: client user or admin.
+   2. `context.req.path`: original request path from client, ex: for login, it is `/auth/login`.
+   3. `context.req.body`: original request body from client, ex: for login, it would be an object with username and password.
+   4. `context.req.id`: original request ID.
+
+## user metadata
+
+For future advanced management requirements, auth gear should have user metadata, it includes common user attributes, such as avatar, first name, last name, display name, preferred language, ..., etc. 
+
+Common user attributes would be great help for better auth gear use experience, which allows to provide API response in preferred language,  multi-lang custom email template.
+
+User metadata is also allowed to store varied user attributes, they differ from application to application, such as: ethnic, height, weight, hobby, ..., etc. It is defined by developer freely.
+
+```
+CREATE TABLE _auth_user_metadata (
+  user_id text REFERENCES _core_user(id),
+  created_at timestamp without time zone NOT NULL,
+  created_by text,
+  updated_at timestamp without time zone NOT NULL,
+  
+  updated_by text,
+  avatar_url text,
+  first_name text,
+  last_name text,
+  display_name text,
+  birthday imestamp without time zone,
+  gender text,
+  prefer_lang_id text REFERENCES _core_lang(id),
+  ...
+  
+  data jsonb,
+  
+  PRIMARY KEY(user_id),
+  UNIQUE (user_id)
+);
+```
+
+The structure of user object could be:
+
+```
+{
+    id: <id>,
+    createdAt: <createdAt>,
```

Comment:
I think here we are showing how the user object looks like only, but not the data transmit? It is important for request and response that we use underscore instead of camelcase for variable names.
For the predefined metadata too, will cloud sdk will do the conversation too? Or we will just keep it simple to use underscore for all platform?